src/rf/apps/workers/process.py

```diff
@@ -367,16 +383,28 @@ def emr_event(self, record, started_status, finished_status):
             return False
 
         if status == FAILED:
-            error = record.get('error', '')
-            return status_updates.update_layer_status(layer_id,
-                                                      enums.STATUS_FAILED,
-                                                      error)
+            default_error = emr_status + ' job failed.'
+            error = record.get('error', default_error)
+            if error.strip() == '':
+                error = default_error
+            return status_updates.mark_layer_status_end(layer_id,
+                                                        emr_status,
+                                                        error)
         elif status == STARTED:
-            return status_updates.update_layer_status(layer_id,
-                                                      started_status)
+            # If chunking started then the cluster creation succeeded.
+            if emr_status == enums.STATUS_CHUNK:
+                status_updates.mark_layer_status_end(
+                    layer_id, enums.STATUS_CREATE_CLUSTER)
+
+            return status_updates.mark_layer_status_start(layer_id, emr_status)
+
         elif status == FINISHED:
-            return status_updates.update_layer_status(layer_id,
-                                                      finished_status)
+            if emr_status == enums.STATUS_MOSAIC:
+                status_updates.mark_layer_status_end(layer_id,
+                                                     enums.STATUS_COMPLETED)
```

Comment:
Consider deleting `emr_event` and moving chunk/mosaic logic into separate methods.